sdk/storage/azure-storage-files-shares/src/share_lease_client.cpp

```diff
@@ -166,6 +166,11 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
           context,
           protocolLayerOptions);
 
+      {
+        std::lock_guard<std::mutex> guard(m_mutex);
```

Comment:
This pattern of locking, getting, release and later lock, set, release is the right pattern - you can't do anything better than this. Godo job!

Reply:
@JeffreyRichter I was wondering if using an std::atomic<std::vector<uint8_t>> could be more efficient to avoid using a mutex.
I was reading about how the OS typically implements the mutex vs atomic, and then the lock is just to read/set a value, which should be very fast, an atomic implementation could be more efficient.  

But as @Jinming-Hu mentioned, it might be trivial customization for this case.

Reply:
I'm fairly certain that atomic will use mutex internally. Check the source of atomic and see.

Reply:
I just looked up https://en.cppreference.com/w/cpp/atomic/atomic/is_lock_free. It says: All atomic types except for std::atomic_flag **_may be implemented using mutexes_** or other locking operations, rather than using the lock-free atomic CPU instructions.

Reply:
More info:
https://en.cppreference.com/w/cpp/atomic/atomic/is_lock_free#:~:text=Atomic%20types%20are%20also%20allowed%20to%20be%20sometimes,of%20the%20same%20type%20have%20to%20use%20locks.

Reply:
Yes, but this only allowed on memory locations that are properly aligned and whose size is accessible by a single CPU instruction. In other words (pun intended), 8, 16, 32, & 64 bit values. Any structure larger than this can't be manipulated by a single CPU instruction and therefore REQUIRES the use of a lock. In this case, we have a std::string which (I'm pretty sure this is larger than 64 bits).

Reply:
Just as an FYI: The ONLY way to synchronize threads running on different cores is to use special CPU instructions - many people call these lock free. But, these instructions only work on values up to 64-bits in size. So, a mutex (and ALL synchronization constructs) are actually built on top of these CPU instructions.
In C++, the atomic type looks at the size (or type) of the T and either uses the CPU instructions (if possible) or uses a mutex (which uses the CPU instructions internally).

The point being that these special CPU instructions are very slow compared to other CPU instructions because they have to coordinate the memory access across ALL the CPU cores and their cache (look up "cache cohernecy" for more info). When taking a mutex that is currently unowned (the common case) - the performance is the same as calling a single CPU instruction. Releasing the mutex calls another CPU instruction. 

Trying to acquire a mutex that is owned has unpredictable performance; in fact, if the owning thread never releases the mutex, the time to acquire a mutex can be infinite time.